### PR TITLE
🛡. Add authentication and authorization

### LINKED ISSUE #173: 🐞  Return 500 Server Errors when items not found
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

Similar to #99 , this is specifically focused for not found items. The error always displays 500 and this is confusing for users of the API. 

### Expected Behavior

_No response_

### Steps To Reproduce

_No response_

### Environment

```markdown
- OS:
```

### Anything else?

_No response_

### LINKED ISSUE #174: 🛡 Add Authentication and Authorization
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Suggestion

Add authentication and authorization. Support JWT bearer tokens and form based auth, along with a few administrative functions to add users and change passwords. Includes examples on the user api's demonstrating how to authorize endpoints based on scopes but it does not add that functionality to all of the endpoints in the api.

### Possible Implementation

_No response_

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by AddressXception**
[//]: # (🚨 Please review the CONTRIBUTING.md in this repository. 💔Thank you!)

### Issue
*Link your PR to an issue*

Fixes #174 
Fixes #173 

### Description
Add authentication and authorization. This PR adds auth support but stubs out the actual auth implementation. Supports JWT bearer tokens and form based auth, along with a few administrative functions to add users and change passwords. It also includes examples on the user api's demonstrating how to authorize endpoints based on scopes but it does not add that functionality to all of the endpoints in the api.

### Testing
1. call `auth/login` with the default credentials
2. verify you can call `/user/me`
3. create a new user as an admin with the `voter` or some other scope
4. login as the new user and try to create a user, expect that it fails (since you aren't an admin)

### CODE REVIEW

File: app/api/v1/auth/auth.py [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+def validate_access_token_authorization(
+    token_data: TokenData, scopes: List[UserScope]
+) -> None:
+    """Validate that the access token is authorized to the requested resource."""
+    if any(scopes):
+        scope_str = ",".join(scopes)
+        authenticate_value = f'Bearer scope="{scope_str}"'
+    else:
+        authenticate_value = "Bearer"
+    for scope in scopes:
```
Comment (AddressXception) [inline app/api/v1/auth/auth.py:69]: this effectively makes scopes an `OR` operation so they can be "admin, or guardian, or auditor..."

File: app/core/settings.py [hunk 2]
```diff
@@ -39,5 +40,11 @@ class Settings(BaseSettings):
     MONGODB_URI: str = "mongodb://root:example@localhost:27017"
     MESSAGEQUEUE_URI: str = "amqp://guest:guest@localhost:5672"
 
+    AUTH_ALGORITHM = "HS256"
+    AUTH_SECRET_KEY = "09d25e094faa6ca2556c818166b7a9563b93f7099f6f0f4caa6cf63b88e8d3e7"
```
Comment (AddressXception) [inline app/core/settings.py:44]: just generated with openssl. not secure.

File: app/main.py [hunk 3]
```diff
@@ -2,14 +2,39 @@
...
 from app.core.settings import Settings
 from app.core.scheduler import get_scheduler
 
+from app.api.v1.models import UserInfo, UserScope
+from app.core import AuthenticationContext, set_auth_credential, set_user_info
+
 logger = getLogger(__name__)
 
 
+def seed_default_user(settings: Settings = Settings()) -> None:
```
Comment (AddressXception) [inline app/main.py:18]: this is not a great solution for seeding the database but it works for now. open to changing it.

**Review by keithrfung — CHANGES_REQUESTED** [review]
Looked this over. Top level this is failing with the linting so that should be addressed first.

File: app/core/settings.py (hunk 2 above)
Comment (keithrfung) [inline app/core/settings.py:44]: This doesn't seem like this should be here.

File: app/core/ballot.py [hunk 4]
```diff
@@ -28,8 +28,8 @@ def get_ballot(
     except Exception as error:
         print(sys.exc_info())
         raise HTTPException(
-            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
-            detail="get ballot failed",
+            status_code=status.HTTP_404_NOT_FOUND,
```
Comment (keithrfung) [inline app/core/ballot.py:31]: These 404 changes feels like a separate PR.

**Comment by keithrfung** [comment]
@SteveMaier-IRT I completed the updates to the linting to get this pull request to pass.

File: app/api/v1/auth/auth.py (hunk 1 above)
Comment (SteveMaier-IRT) [reply app/api/v1/auth/auth.py:69]: I thought we were looking at having the ability to have someone as both an admin and a guardian.

File: app/core/ballot.py (hunk 4 above)
Comment (SteveMaier-IRT) [reply app/core/ballot.py:40]: A new issue was added to cover this.

File: app/core/settings.py (hunk 2 above)
Comment (SteveMaier-IRT) [reply app/core/settings.py:51]: This and other private setting will need to be moved before we deploy.

File: app/core/ballot.py (hunk 4 above)
Comment (SteveMaier-IRT) [reply app/core/ballot.py:40]: new issue was written for these

File: app/main.py (hunk 3 above)
Comment (SteveMaier-IRT) [reply app/main.py:18]: Could use secret env settings for the setup.  we can talk about it.

**Review by SteveMaier-IRT — APPROVED** [review]

File: app/api/v1/auth/auth.py (hunk 1 above)
Comment (AddressXception) [reply app/api/v1/auth/auth.py:69]: yes. this does that. sorry wording not clear. i mean either or can access, not that they are "or exclusive"

**Pull request merged**